drivers/clock_control/clock_control_rcar_cpg_mssr.c

```diff
@@ -56,76 +26,60 @@ static void cpg_write(const struct rcar_mssr_config *config,
 	k_sleep(K_USEC(35));
 }
 
-static void cpg_reset(const struct rcar_mssr_config *config,
-		      uint32_t reg, uint32_t bit)
-{
-	cpg_write(config, SRCR(reg), BIT(bit));
-	cpg_write(config, SRSTCLR(reg), BIT(bit));
-}
-
-static int cpg_core_clock_endisable(const struct device *dev,
-				    uint32_t module, uint32_t rate, bool enable)
+static int cpg_reset(const struct rcar_mssr_config *config,
+		     uint32_t reg, uint32_t bit)
 {
-	const struct rcar_mssr_config *config = dev->config;
-	uint32_t divider;
-	unsigned int key;
+	uint16_t offset;
 	int ret;
 
-	/* Only support CANFD core clock at the moment */
-	if (module != CPG_CORE_CLK_CANFD) {
-		return -EINVAL;
+	ret = soc_cpg_get_srcr_offset(reg, &offset);
+	if (ret < 0) {
+		return ret;
 	}
 
-	key = irq_lock();
-
-	if (enable) {
-		if ((CANFDCKCR_PARENT_CLK_RATE % rate) != 0) {
-			__ASSERT(true, "Can not generate "
-				 "%u from CANFD parent clock", rate);
-			ret = -EINVAL;
-			goto unlock;
-		}
-		divider = (CANFDCKCR_PARENT_CLK_RATE / rate) - 1;
-		if (divider > CANFDCKCR_DIVIDER_MASK) {
-			__ASSERT(true, "Can not generate %u "
-				 "from CANFD parent clock", rate);
-			ret = -EINVAL;
-			goto unlock;
-		}
-		cpg_write(config, CANFDCKCR, divider);
-	} else {
-		cpg_write(config, CANFDCKCR, CANFDCKCR_CKSTP);
-	}
+	cpg_write(config, offset, BIT(bit));
+	cpg_write(config, SRSTCLR(reg), BIT(bit));
 
-unlock:
-	irq_unlock(key);
 	return 0;
 }
 
-static int cpg_rmstp_clock_endisable(const struct device *dev,
-				     uint32_t module, bool enable)
+static int cpg_mstp_clock_endisable(const struct device *dev,
+				    uint32_t module, bool enable)
 {
 	const struct rcar_mssr_config *config = dev->config;
 	uint32_t reg = module / 100;
 	uint32_t bit = module % 100;
 	uint32_t bitmask = BIT(bit);
 	uint32_t reg_val;
+	uint16_t offset;
+	int ret;
+	unsigned int key;
 
-	__ASSERT((bit < 32) && reg < ARRAY_SIZE(rmstpsr),
-		 "Invalid module number for cpg clock: %d", module);
+	if (bit > 32) {
+		return -EINVAL;
+	}
 
-	unsigned int key = irq_lock();
+	ret = soc_cpg_get_mstpcr_offset(reg, &offset);
+	if (ret < 0) {
+		return ret;
+	}
 
-	reg_val = sys_read32(config->base_address + RMSTPSR(reg));
+	key = irq_lock();
+
+	reg_val = sys_read32(config->base_address + offset);
 	if (enable) {
 		reg_val &= ~bitmask;
 	} else {
 		reg_val |= bitmask;
 	}
 
-	sys_write32(reg_val, config->base_address + RMSTPSR(reg));
+	sys_write32(reg_val, config->base_address + offset);
 	if (!enable) {
-		cpg_reset(config, reg, bit);
+		ret = cpg_reset(config, reg, bit);
+		if (ret < 0) {
```

Comment:
you can initialise ret to 0
and just return ret at the end of the function, so that you don't have to call irq_unlock

Reply:
As we are returning from this method if `soc_cpg_get_mstpcr_offset` is returning an error, ret should still be at 0 if we reach `cpg_reset call`, so it's a good way to clean the code a bit. thanks